File: client/protocol_v1.js
"use strict";

const ALLOWED_TYPES = new Set(["MSG", "RESPONSE", "ERROR"]);

function encodeFrame(type, sequence, payload) {
  if (!ALLOWED_TYPES.has(type)) throw new Error(`unsupported type: ${type}`);
  if (!Number.isInteger(sequence) || sequence < 0) throw new Error("sequence must be a non-negative integer");

  const payloadBytes = Buffer.from(payload, "utf8");
  return Buffer.concat([Buffer.from(`V1|${type}|${sequence}|${payloadBytes.length}|`, "ascii"), payloadBytes]);
}

function decodeFrame(raw) {
  const bytes = Buffer.from(raw);
  const separators = [];
  for (let index = 0; index < bytes.length && separators.length < 4; index += 1) {
    if (bytes[index] === 124) separators.push(index);
  }
  if (separators.length !== 4) throw new Error("frame must contain five fields");

  const version = bytes.subarray(0, separators[0]).toString("ascii");
  const type = bytes.subarray(separators[0] + 1, separators[1]).toString("ascii");
  const sequence = Number(bytes.subarray(separators[1] + 1, separators[2]).toString("ascii"));
  const declaredLength = Number(bytes.subarray(separators[2] + 1, separators[3]).toString("ascii"));
  const payloadBytes = bytes.subarray(separators[3] + 1);

  if (version !== "V1") throw new Error(`unsupported version: ${version}`);
  if (!ALLOWED_TYPES.has(type)) throw new Error(`unsupported type: ${type}`);
  if (!Number.isInteger(sequence) || sequence < 0 || !Number.isInteger(declaredLength) || declaredLength < 0) {
    throw new Error("sequence and length must be non-negative integers");
  }
  if (payloadBytes.length !== declaredLength) throw new Error("declared payload length does not match datagram");

  return { version, type, sequence, payload: payloadBytes.toString("utf8") };
}

module.exports = { decodeFrame, encodeFrame };

File: server/protocol_v1.py
"""Wire format shared by the V1 server implementation.

Frames are UTF-8 bytes with the format: V1|TYPE|SEQUENCE|LENGTH|PAYLOAD.
LENGTH is the number of UTF-8 bytes in PAYLOAD.
"""

from __future__ import annotations


ALLOWED_TYPES = {"MSG", "RESPONSE", "ERROR"}


class FrameError(ValueError):
    """Raised when a datagram does not match the V1 frame format."""


def encode_frame(message_type: str, sequence: int, payload: str) -> bytes:
    if message_type not in ALLOWED_TYPES:
        raise FrameError(f"unsupported type: {message_type}")
    if not isinstance(sequence, int) or sequence < 0:
        raise FrameError("sequence must be a non-negative integer")

    payload_bytes = payload.encode("utf-8")
    header = f"V1|{message_type}|{sequence}|{len(payload_bytes)}|"
    return header.encode("ascii") + payload_bytes


def decode_frame(raw: bytes) -> dict[str, object]:
    fields = raw.split(b"|", 4)
    if len(fields) != 5:
        raise FrameError("frame must contain five fields")
    header, payload_bytes = fields[:4], fields[4]

    try:
        version, message_type, sequence_text, length_text = [field.decode("ascii") for field in header]
        sequence = int(sequence_text)
        declared_length = int(length_text)
        payload = payload_bytes.decode("utf-8")
    except (UnicodeDecodeError, ValueError) as error:
        raise FrameError("frame header or payload is invalid") from error

    if version != "V1":
        raise FrameError(f"unsupported version: {version}")
    if message_type not in ALLOWED_TYPES:
        raise FrameError(f"unsupported type: {message_type}")
    if sequence < 0 or declared_length < 0:
        raise FrameError("sequence and length must be non-negative")
    if declared_length != len(payload_bytes):
        raise FrameError("declared payload length does not match datagram")

    return {"version": version, "type": message_type, "sequence": sequence, "payload": payload}
